### PR TITLE
fix message passing

### PR DESCRIPTION
**Pull request by chhsiao93**
# RFC: Fix Edge Feature Processing in InteractionNetwork

## Summary
The current implementation of `InteractionNetwork` does not properly propagate processed edge features through the message passing flow, causing the residual connection to simply double the original edge features instead of adding processed features to the original ones.

## Background
In the current implementation, the `message()` method processes edge features but these processed features are not properly passed through to the `update()` method. This causes the `update()` method to use the original edge features instead of the processed ones, resulting in incorrect residual connections.

## Problem
When running the network:
```python
edge_attr = tensor([[1., 1.],
                   [2., 2.]])
```
produces:
```python
output = tensor([[2., 2.],
                [4., 4.]])
```
This indicates the output is simply doubling the input due to the residual connection, rather than adding processed features to the original ones.

## Proposed Solution
Add a temporary instance variable to store processed edge features between message and update steps:

```python
def message(self,
           x_i: torch.tensor,
           x_j: torch.tensor,
           edge_features: torch.tensor) -> torch.tensor:
    edge_features = torch.cat([x_i, x_j, edge_features], dim=-1)
    self._edge_features = self.edge_fn(edge_features)  # Store processed features
    return self._edge_features  # Return for node updates

def update(self,
          x_updated: torch.tensor,
          x: torch.tensor,
          edge_features: torch.tensor):
    x_updated = torch.cat([x_updated, x], dim=-1)
    x_updated = self.node_fn(x_updated)
    return x_updated, self._edge_features  # Use stored processed features
```

## Rationale
- No need to modify the PyTorch Geometric message passing interface
- Maintains proper feature processing flow
- Clean and simple solution that follows PyG patterns
- No performance overhead as the variable only exists during forward pass

## Testing
New tests verify:
1. Edge features are properly updated (not just doubled)
2. Shapes remain correct
3. Gradients flow properly
4. Numerical stability is maintained

## Alternatives Considered
1. Using class member variable initialized in `__init__`
   - Rejected: Unnecessary persistence between forward passes
2. Modifying PyG's message passing flow
   - Rejected: Too invasive, breaks standard patterns
3. Using message dictionaries
   - Rejected: Not compatible with PyG's aggregation

## Backward Compatibility
This change maintains the same interface and output shapes, but will produce different numerical results. Models trained with the old implementation will need to be retrained.

### CODE REVIEW

File: gns/graph_network.py [hunk 1]
```diff
@@ -219,7 +220,7 @@ def update(self,
     # [nparticles, latent_dim (or nnode_in) *2]
```
Comment (kks32) [inline gns/graph_network.py:220]:
```suggestion
    # [nparticles, latent_dim (or nnode_in) *2]
    # This gets called later, after message() and aggregate()
    # Update modified from MessagePassing takes the output of aggregation
    # as first argument and any argument which was initially passed to
    # propagate hence we need to return the stored value of edge_features
```

File: gns/graph_network.py [hunk 2]
```diff
@@ -195,8 +195,9 @@ def message(self,
     """
     # Concat edge features with a final shape of [nedges, latent_dim*3]
     edge_features = torch.cat([x_i, x_j, edge_features], dim=-1)
-    edge_features = self.edge_fn(edge_features)
-    return edge_features
+    self._edge_features  = self.edge_fn(edge_features)
```
Comment (kks32) [inline gns/graph_network.py:198]:
```suggestion
    self._edge_features = self.edge_fn(edge_features)  # Create and store
    return self._edge_features  # This gets passed to aggregate()
```

**Review by kks32 — CHANGES_REQUESTED** [review]
Add the following to forward() function:
```
        # Start propagating messages.
        # Takes in the edge indices and all additional data which is needed to
        # construct messages and to update node embeddings.
        # Call PyG propagate() method:
        # 1. Message phase - compute messages for each edge
        # 2. Aggregate phase - aggregate messages for each node
        # 3. Update phase - updates only the node features
        # Update uses the message from step 1 and any original arguments passed to 
        # propagate() to update the node embeddings. This is why we need to store
        # the updated edge features to return them from the update() method.
```

can you run black --check

**Comment by kks32** [comment]
@chhsiao93 Could you add this fix to v2.0 branch too?

**Review by kks32 — APPROVED** [review]
